File: hapi-fhir-demo/src/main/java/com/demo/fhir/auth/service/AuthService.java
package com.demo.fhir.auth.service;

import com.demo.fhir.auth.dto.LoginRequest;
import com.demo.fhir.auth.dto.LoginResponse;
import com.demo.fhir.auth.dto.RegisterRequest;
import com.demo.fhir.auth.model.AppUser;
import com.demo.fhir.auth.repository.AuthUserRepository;
import com.demo.fhir.shared.security.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;
import java.util.UUID;

@Service
public class AuthService {

    @Autowired
    private AuthUserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    @Value("${jwt.access-token-expiry-seconds}")
    private long accessTokenExpiry;

    @Value("${jwt.refresh-token-expiry-seconds}")
    private long refreshTokenExpiry;

    // ── Register ──────────────────────────────────────────────────────────────

    @Transactional
    public AppUser register(RegisterRequest request) {
        if (userRepository.findByUsername(request.getUsername()).isPresent()) {
            throw new IllegalArgumentException(
                    "Username already taken: " + request.getUsername());
        }

        AppUser user = new AppUser();
        user.setUsername(request.getUsername());
        user.setPasswordHash(passwordEncoder.encode(request.getPassword()));
        user.setRole(request.getRole());
        user.setPatientId(request.getPatientId());

        return userRepository.save(user);
    }

    // ── Login ─────────────────────────────────────────────────────────────────

    @Transactional
    public LoginResponse login(LoginRequest request) {
        AppUser user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() ->
                        new IllegalArgumentException("Invalid username or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new IllegalArgumentException("Invalid username or password");
        }

        String accessToken  = buildAccessToken(user);
        String refreshToken = buildAndStoreRefreshToken(user);

        return new LoginResponse(accessToken, refreshToken,
                                 user.getUsername(), user.getRole());
    }

    // ── Refresh ───────────────────────────────────────────────────────────────

    @Transactional
    public String refresh(String rawRefreshToken) {
        // Parse to get username — throws JwtException if expired/invalid
        var claims  = jwtUtil.parse(rawRefreshToken);
        String type = claims.get("type", String.class);
        if (!"refresh".equals(type)) {
            throw new IllegalArgumentException("Not a refresh token");
        }

        String username = claims.getSubject();
        AppUser user = userRepository.findByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        // Verify the hashed token matches what we stored
        if (user.getRefreshTokenHash() == null
                || !passwordEncoder.matches(rawRefreshToken, user.getRefreshTokenHash())) {
            throw new IllegalArgumentException("Refresh token has been invalidated");
        }

        return buildAccessToken(user);
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private String buildAccessToken(AppUser user) {
        Map<String, Object> claims = Map.of(
                "sub",       user.getUsername(),
                "role",      user.getRole().name(),
                "patientId", user.getPatientId() != null ? user.getPatientId() : "",
                "type",      "access"
        );
        return jwtUtil.sign(claims, accessTokenExpiry);
    }

    private String buildAndStoreRefreshToken(AppUser user) {
        String jti = UUID.randomUUID().toString();
        Map<String, Object> claims = Map.of(
                "sub",  user.getUsername(),
                "role", user.getRole().name(),
                "jti",  jti,
                "type", "refresh"
        );
        String rawToken = jwtUtil.sign(claims, refreshTokenExpiry);
        // Store the hash — never store the raw token in the DB
        user.setRefreshTokenHash(passwordEncoder.encode(rawToken));
        userRepository.save(user);
        return rawToken;
    }
}

File: frontend/src/services/authService.js
import api from './api';

// ─── Mock Data Fallback ──────────────────────────────────────────────────────
const MOCK_ENABLED = false;

const MOCK_LOGIN = {
  accessToken: 'mock-access-token-abc123',
  refreshToken: 'mock-refresh-token-xyz789',
  username: 'demo_user',
  role: 'DOCTOR',
};

export const authService = {
  login: async (username, password, role) => {
    try {
      const res = await api.post('/auth/login', { username, password });
      return res.data;
    } catch (err) {
      if (MOCK_ENABLED) {
        console.warn('[Mock] Login fallback active');
        return { ...MOCK_LOGIN, role: role || 'DOCTOR', username };
      }
      throw err;
    }
  },

  register: async (username, password, role, patientId) => {
    try {
      const res = await api.post('/auth/register', {
        username,
        password,
        role,
        ...(role === 'PATIENT' && patientId ? { patientId } : {}),
      });
      return res.data;
    } catch (err) {
      if (MOCK_ENABLED) {
        console.warn('[Mock] Register fallback active');
        return { message: 'User registered successfully', username, role };
      }
      throw err;
    }
  },

  refresh: async (refreshToken) => {
    const res = await api.post('/auth/refresh', { refreshToken });
    return res.data;
  },
};
